src/instantsend/db.cpp

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) 2019-2025 The Dash Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <instantsend/db.h>
+
+#include <chain.h>
+#include <dbwrapper.h>
+#include <primitives/block.h>
+#include <util/system.h>
+
+static const std::string_view DB_ARCHIVED_BY_HASH = "is_a2";
+static const std::string_view DB_ARCHIVED_BY_HEIGHT_AND_HASH = "is_a1";
+static const std::string_view DB_HASH_BY_OUTPOINT = "is_in";
+static const std::string_view DB_HASH_BY_TXID = "is_tx";
+static const std::string_view DB_ISLOCK_BY_HASH = "is_i";
+static const std::string_view DB_MINED_BY_HEIGHT_AND_HASH = "is_m";
+static const std::string_view DB_VERSION = "is_v";
+
+namespace instantsend {
+namespace {
+static std::tuple<std::string, uint32_t, uint256> BuildInversedISLockKey(std::string_view k, int nHeight, const uint256& islockHash)
+{
+    return std::make_tuple(std::string{k}, htobe32_internal(std::numeric_limits<uint32_t>::max() - nHeight), islockHash);
+}
+} // anonymous namespace
+
+CInstantSendDb::CInstantSendDb(bool unitTests, bool fWipe) :
+    db(std::make_unique<CDBWrapper>(unitTests ? "" : (gArgs.GetDataDirNet() / "llmq/isdb"), 32 << 20, unitTests, fWipe))
+{
+    Upgrade(unitTests);
+}
+
+CInstantSendDb::~CInstantSendDb() = default;
+
+void CInstantSendDb::Upgrade(bool unitTests)
+{
+    LOCK(cs_db);
+    int v{0};
+    if (!db->Read(DB_VERSION, v) || v < CInstantSendDb::CURRENT_VERSION) {
+        // Wipe db
+        db.reset();
+        db = std::make_unique<CDBWrapper>(unitTests ? "" : (gArgs.GetDataDirNet() / "llmq/isdb"), 32 << 20, unitTests,
+                                          /*fWipe=*/true);
+        CDBBatch batch(*db);
+        batch.Write(DB_VERSION, CInstantSendDb::CURRENT_VERSION);
+        // Sync DB changes to disk
+        db->WriteBatch(batch, /*fSync=*/true);
+        batch.Clear();
+    }
+}
+
+void CInstantSendDb::WriteNewInstantSendLock(const uint256& hash, const InstantSendLock& islock)
+{
+    LOCK(cs_db);
+    CDBBatch batch(*db);
+    batch.Write(std::make_tuple(DB_ISLOCK_BY_HASH, hash), islock);
+    batch.Write(std::make_tuple(DB_HASH_BY_TXID, islock.txid), hash);
+    for (const auto& in : islock.inputs) {
+        batch.Write(std::make_tuple(DB_HASH_BY_OUTPOINT, in), hash);
+    }
+    db->WriteBatch(batch);
+
+    islockCache.insert(hash, std::make_shared<InstantSendLock>(islock));
+    txidCache.insert(islock.txid, hash);
+    for (const auto& in : islock.inputs) {
+        outpointCache.insert(in, hash);
+    }
+}
+
+void CInstantSendDb::RemoveInstantSendLock(CDBBatch& batch, const uint256& hash, InstantSendLockPtr islock, bool keep_cache)
```

Comment:
nit not for this pr

@coderabbitai, can this islock be taken in as a raw pointer to avoid having to copy in the shared_ptr?